NPC support with Citizens

plexus opened the pull request:
Ongoing effort to develop a good story around supporting [Citizens 2](https://wiki.citizensnpcs.co/Citizens_Wiki)

plexus committed:
Citizens: find a better solution for traits

devurandom commented on src/lambdaisland/witchcraft/citizens/trait.clj:
```diff
@@ -0,0 +1,123 @@
+  that you declare in your `witchcraft.edn` plugin config should do the trick.
+
+  To make this work we `gen-class` a concrete subclass, but instead of writing
+  it out to we just compile and load it in memory (DynamicClassLoader for the
```
```suggestion
  it out to a file we just compile and load it in memory (DynamicClassLoader for the
```

devurandom commented on src/lambdaisland/witchcraft/citizens/trait.clj:
```diff
@@ -0,0 +1,123 @@
+                                       TraitInfo)
+           (org.bukkit.entity EntityType)))
+
+(defn trait-factory ^TraitFactory
+  "Citizens' TraitFactory"
+  []
+  (CitizensAPI/getTraitFactory))
```
I noticed some duplications with the `citizens` namespace. Is that intentional?

plexus replied:
Yeah for now that's intentional

devurandom commented on src/lambdaisland/witchcraft/citizens.clj:
```diff
@@ -0,0 +1,63 @@
+               (get wc/entity-types entity-type)
+               npc-name))
+
+(defn make-trait
+  {:doc (:doc (meta #'trait/make-trait))}
+  ^Class [name callbacks]
+  (trait/make-trait name callbacks))
```
This seems to just copy a function from the `trait` namespace, which is also required and thus accessible. What is the reason for not using this function directly?

plexus replied:
It's just a matter of exposing the API, the trait namespace can be considered private/implementation detail, I may just roll it into the citizens namespace entirely.

devurandom commented on src/lambdaisland/witchcraft/citizens/trait.clj:
```diff
@@ -0,0 +1,123 @@
+  []
+  (CitizensAPI/getTraitFactory))
+
+(defn gen-and-load-class
+  "Generates and immediately loads the bytecode for the specified
+  class. Note that a class generated this way can be loaded only once
+  - the JVM supports only one class with a given name per
+  classloader. Subsequent to generation you can import it into any
+  desired namespaces just like any other class. See gen-class for a
+  description of the options.
+
+  Taken from the rich comment section of `clojure/genclass.clj`.
+  "
+  [options-map]
+  (let [[cname bytecode] (#'clojure.core/generate-class options-map)]
+    (.defineClass (licp/root-loader)
+                  (str/replace cname
+                               #"/"
+                               ".")
+                  bytecode
+                  "")))
```
This seems to be a utility function of general value. Is there maybe a better place for this?

plexus replied:
It could move to util if we end up using it elsewhere. Not a concern right now.

devurandom commented on src/lambdaisland/witchcraft/citizens/trait.clj:
```diff
@@ -0,0 +1,123 @@
+                  "")))
+
+(defonce
+  ^{:doc "Registry from trait name to mape of callbacks"}
```
```suggestion
  ^{:doc "Registry from trait name to map of callbacks"}
```

devurandom commented on src/lambdaisland/witchcraft/citizens/trait.clj:
```diff
@@ -0,0 +1,123 @@
+          :init init-fn-name
+          :constructors {[] [String]}
+          ;;:post-init `my-post-init
+          :prefix "trait-"
+          ;;:impl-ns "my.trait.impl"
```
This looks for the implementation of methods in the current namespace (`lambdaisland.witchcraft.citizens.trait`) with the always-same prefix `trait-`, right? So there can be only one trait globally? (And even if `:impl-ns` was set, there could be only one trait per namespace?)

Parts of this appear to have to be moved into a macro to become more generally useful (so that the namespace of the caller can be passed into this function, so that the implementation is looked for in a namespace the user has access to). But it is equally possible that I am missing something here. Please explain.

plexus replied:
All of our trait subclasses will share the same method implementations, but these implementations simply delegate to the trait-specific callbacks in the registry. I do it this way because reloading classes/class methods when they change is not possible (or at least a huge pain). The extra indirection gives us the possibility to easily redefine things.

devurandom replied:
I see, that is why you go via `call-cb` and `registry`. Thanks!

devurandom commented on src/lambdaisland/witchcraft/citizens.clj:
```diff
@@ -0,0 +1,63 @@
+(defn trait-factory ^TraitFactory []
+  (CitizensAPI/getTraitFactory))
+
+(defprotocol HasCreateNPC
+  (-create-npc [_ e s]))
+
+(reflect/extend-signatures HasCreateNPC
+                           "createNPC(org.bukkit.entity.EntityType,java.lang.String)"
+                           (-create-npc [this entity-type npc-name]
+                                        (.createNPC this entity-type npc-name)))
+
+(defn create-npc [entity-type npc-name]
+  (-create-npc (npc-registry)
+               (get wc/entity-types entity-type)
+               npc-name))
```
Why is this necessary? What does it do?

I currently read this as "marking" all classes on the classpath that have a `createNPC` method as implementing the `HasCreateNPC` protocol, by "wrapping" the `createNPC` Java method in a `-create-npc` Clojure function. Is this correct? Why is this done? (Because there is no Java interface defined and implemented by Citizens? And if so, why would we need one?) Why can `createNPC` not be called directly?

plexus replied:
This is generally how we implement interop in Witchcraft (see also the many calls to extend-signature in the main namespace). It's basically a form of "duck typing" (https://en.wikipedia.org/wiki/Duck_typing), we want an API that can work with any type of object that has the required behavior, regardless of the concrete types. This mechanism is also what allows us to be compatible with multiple versions of servers and versions of minecraft. We don't program against concrete types, we program against protocols, and wire those up based on what we detect on the classpath.

devurandom replied:
Just for my own understanding: Is this strictly necessary here or did you add it because it will come in handy later? `npc-registry` has a type annotation, so I would expect that the Clojure compiler knows that a `createNPC` method with 2 arguments (+ this) exists, so it does not have to use reflection to find it when calling it on `(npc-registry)`.

plexus replied:
In this case it's not strictly necessary, no.

devurandom reviewed (commented):
Thanks a lot for this! It is generally similar to the code I played with locally when trying out Citizens, but extends it a lot. I also like the documentation. Having this in Witchcraft will be great!

I have a bunch of questions. I am not very familiar with the techniques of dynamic code generation and Clojure/Java interop used here.

plexus committed:
Make our traits stateful, and honor load/save

This provides a mechanism for Clojure-based trait implementations to carry state
by making the Trait instance quack like an Atom. It implements default load/save
methods that handle persisting this state, so that this is handled
transparently (see `plugins/Citizens/saves.yml`).

This introduces two new values for the options map to `make-trait`: `:init` and
`:post-init`. `:init` is a map or zero-arg function returning a map, which is
used to set the initial state when instantiating a new trait. `:post-init`
corresponds with `gen-class`'s `:post-init`, it's a single-arg function that
receives the instantiated Trait object.

devurandom commented on src/lambdaisland/witchcraft/citizens.clj:
```diff
@@ -42,7 +52,10 @@
-  (.getOrAddTrait npc (trait-class trait-name)))
+  (-get-or-add-trait npc (trait-class trait-name)))
+
+(defn npc-by-id [id]
+  (.getById (npc-registry) id))
```
Q: For `.createNPC` you created a protocol. Should this one also have one?

devurandom commented on src/lambdaisland/witchcraft/citizens/trait.clj:
```diff
@@ -0,0 +1,228 @@
+          :state "state"
+          :init init-fn-name
+          :constructors {[] [String]}
+          :post-init `trait-post-init
```
```suggestion
          :post-init "post-init"
```

Otherwise I get:
```
Execution error (ClassFormatError) at java.lang.ClassLoader/defineClass1 (ClassLoader.java:-2).
Illegal field name "lambdaisland.witchcraft.citizens.trait_SLASH_trait_post_init__var" in class my/name/space
```
when calling `make-trait`.

plexus replied:
Yeah sorry I hadn't pushed my code in a while. I pushed my latest changes, including two example implementations of traits.

plexus committed:
More witchcraft helpers

- entity
- get/set difficulty
- inventory stuff

plexus committed:
More work on Citizens supports

Implement persisted state for traits, and more helpers

plexus committed:
Add example of a trait implementation

plexus committed:
Add example of a trait implementation

plexus commented:
This is starting to work nicely, but I'm still fine tuning the persistence of traits.
